CONTRIBUTING.md

```diff
@@ -1,25 +1,74 @@
 
-Create an issue with a reproducible example.  Please specify the Rust version, delta-kernel-rs version, the code executed, and the error message.
+## How to Contribute
 
-## Did you create a PR to fix a bug?
+We follow a structured contribution process to ensure high-quality code:
 
-Open a pull request and add "Fixes #issue_number" in the PR description.
+1. **Start with an issue and/or design sketch**: Open an issue describing what you want to
+   contribute and why. Continue to step 2 after reaching some consensus. This helps us avoid wasted
+   effort (perhaps you were building something that someone else was already pursuing or already
+   explored and rejected). Including a design sketch will help drive consensus (often a simple
+   diagram or bullet points outlining high-level changes is sufficient).
+2. **Prototype/POC**: Create a PR marked as prototype/draft (not intended to merge) and gather
+   feedback to further derisk the design. This PR is not intended to be merged but will guide the
+   implementation and serve as a proving ground for the design. Then, pieces are torn out into
+   smaller PRs that can be merged.
+3. **Implementation**: Finally, create PR(s) to implement the feature (production code, tests,
+   thorough docs, etc.). Often the initial POC will be split into multiple smaller PRs (e.g.,
+   refactors, then feature additions, then public APIs specifically). Care should be taken to ensure
+   each PR is easily reviewable and thoroughly (independently) tested.
 
-We appreciate bug fixes - thank you in advance!
+## Forking and Setup
 
-## Would you like to add a new feature or change existing code?
+1. Fork the repository into your account
+2. Clone your fork locally:
+   ```bash
+   git clone git@github.com:YOUR_USERNAME/delta-kernel-rs.git
+   cd delta-kernel-rs
+   ```
+3. Add the upstream remote:
+   ```bash
+   git remote add upstream git@github.com:delta-io/delta-kernel-rs.git
+   ```
 
-If you would like to add a feature or change existing behavior, please make sure to create an issue and get the planned work approved by the core team first!
+Now you have:
+- `origin` pointing to your fork
+- `upstream` pointing to the original repository
 
-Always better to get aligned with the core devs before writing any code.
+## Development Workflow
 
-## Do you have questions about the source code?
+Our trunk branch is named `main`. Here's the typical workflow:
 
-Feel free to create an issue or join the [Delta Lake Slack](https://go.delta.io/slack) with questions!  We chat in the `#delta-kernel` channel.
+1. Pull the latest main to get a fresh checkout:
+   ```bash
+   git checkout main
+   git pull upstream main
+   ```
+2. Create a new feature branch:
+   ```bash
+   git checkout -b my-feature
+   ```
+   (NB: Consider using `git worktrees` for managing multiple branches!)
+3. Make your changes and test them (cargo test, run clippy, etc. - see what our CI runs)
```

Comment:
Let's include the actual commands 
```suggestion
3. Make your changes and test them using the following commands. See our CI runs for a full set of tests.
    ```rust
     cargo test --workspace --verbose --all-features -- --skip read_table_version_hdfs
     cargo clippy --benches --tests --all-features -- -D warnings`
      cargo doc --workspace --all-features
    ```
```